src/mlir/lowering.md

```diff
@@ -0,0 +1,517 @@
+previously declared (in program order).
+
+Furthermore, union types may have compacting optimisations and discriminators to
+match which value is stored, so will need non-trivial extraction lowering (mask
+and shift on the discriminator, then mask, shift and possibly extend on the
+payload). Writes will have to do the reverse process to save values in the ABI
+representation.
+
+### <Integer|Float>LiteralStmt
+
+Constant literals with generic type (int, float). The type we assign to them
+will depend on how they're used and therefore will depend on type inference and
+reachability analysis, which should have been completed by the time the 1IR is
+ready to be converted in the 2IR.
+
+Literals are lowered as immutable objects of the specified concrete type,
+initialised with the literal value.
+
+Example:
+```
+  // 10 + a : U32
+  %const = verona.constant "10" : U32
+  %0 = verona.new_object @U32 [ %const ]
+  %1 = verona.call "+" [ %0 ] ( %a )
+```
+
+### StringLiteralStmt
+
+Strings are complicated on both Verona and MLIR.
+
+Verona's model is converging to one where strings can have different
+representations but a unifying interface, so that concrete string types can
+coexist and be converted to one another. This adds lowering as well as
+representation complexities.
+
+Furthermore, strings can be interpolated with values. If the values are
+compile-time constants, then their string representation can be expanded at
+compile-time. If not, we need to splice the string into a list of literal +
+value + literal, etc. We may need to do that _before_ the concrete representation
+is chosen.
+
+MLIR doesn't support strings at all. We'll have to keep their values in some meta
+storage until we can lower them into LLVM. LLVM IR supports string literals and
+we can call string maniupulation library functions to operate on them. The MLIR
+Toy example has a mock implementation of that process.
+
+Example:
+```
+  // Simple string
+  %simple_string = verona.constant "This is a string" : !verona.string
+
+  // Interpolated string ("My name is $1 and I'm $2 years old", name, age)
+  %0 = verona.constant "My name is " : !verona.string
+  %1 = verona.constant " and I'm " : !verona.string
+  %2 = verona.constant " years old" : !verona.string
+
+  // This is still not decided, but just for illustration purposes
+  // We could also pass this to the representation, so it can do its own thing
+  %name_str = verona.call "toString" [ %name ] : !verona.string
+  %age_str = verona.call "toString" [ %age ] : !verona.string
+
+  // This is whatever runtime string representation we choose
+  // assuming we have a constructor with a list of literal strings
+  %utf8 = verona.new_object @UTF8String [%0, %name_str, %1, %age_str, %2]
+```
+
+In the case above, if all variables are already compile-time constant strings
+at the time of lowering, the compiler has the ability to fuse the strings,
+creating a simple string (or at least less splits).
+
+### UnitStmt
+
+No idea.
```

Comment:
In the long term `unit` won't be a special concept in the IR and will be replaced with something from the standard library.